src/Utils/Hooks/useImagePerformanceObserver.ts

```diff
@@ -0,0 +1,93 @@
+import { useEffect, useRef } from "react"
+import { sendToVolley, VolleyMetric } from "Server/volley"
+import { getENV } from "Utils/getENV"
+import { GEMINI_CLOUDFRONT_URL } from "Utils/resizer"
+
+const DEVICE_TYPE = getENV("IS_MOBILE") ? "mobile" : "desktop"
+
+export const useImagePerformanceObserver = () => {
+  const queue = useRef<
+    Pick<
+      PerformanceResourceTiming,
+      "initiatorType" | "transferSize" | "name" | "duration"
+    >[]
+  >([])
+
+  useEffect(() => {
+    if (typeof PerformanceObserver === "undefined") return
+
+    const performanceObserver = new PerformanceObserver(observedEntries => {
+      const entries = observedEntries.getEntries()
+
+      entries.forEach((entry: PerformanceResourceTiming) => {
+        if (
+          // Only log images
+          entry.initiatorType !== "img" ||
+          // Ensure they are uncached
+          entry.transferSize === 0 ||
+          // Ensure they are Gemini images
+          !entry.name.includes(GEMINI_CLOUDFRONT_URL)
+        ) {
+          return
+        }
+
+        queue.current.push(entry)
+      })
+    })
+
+    performanceObserver.observe({
+      type: "resource",
+      buffered: true,
+    })
+
+    return () => {
+      performanceObserver.disconnect()
+    }
+  }, [])
+
+  // Every 5 seconds, send the queue to the server
+  useEffect(() => {
+    const interval = setInterval(() => {
+      const entries = queue.current
+
+      if (entries.length === 0) return
+
+      const data = entries.map(entry => {
+        const payload: VolleyMetric = {
+          type: "timing",
+          name: "image.load-time",
+          timing: entry.duration,
+          tags: [
+            `transfer-size:${bucket(entry.transferSize)}`,
+            `device-type:${DEVICE_TYPE}`,
+          ],
+        }
+
+        return payload
+      })
+
+      sendToVolley(data)
+
+      queue.current = []
+    }, 5000)
+
+    return () => {
```

Comment:
interesting! i have not seen this before, was looking at https://reactjs.org/docs/hooks-effect.html

> Why did we return a function from our effect? This is the optional cleanup mechanism for effects. Every effect may return a function that cleans up after it. This lets us keep the logic for adding and removing subscriptions close to each other. They’re part of the same effect!

Thinking about all the `useEffect` i have written w/o knowing about/doing any clean-up. At least for the popcorn bidding websockets I think that's likely a good addition?!

the more you know ©️

Reply:
Oh yeah, if you open a socket you'd want to close it!

Reply:
I guess the cleanup in `useEffect` is only needed if there's something to cleanup, which isn't all the time? Makes sense now that i think about it for things like `setTimer`, maybe opening a websocket or other global-ish thing, binding some event handlers. what other clean-ups would one typically do?

Reply:
Yep, exactly those cases. I don't know, tracking the leave state of a component?